File: ticktock-compiler/src/main/kotlin/dev/zacsweers/ticktock/compiler/KotlinWriter.java
package dev.zacsweers.ticktock.compiler;

import com.squareup.kotlinpoet.CodeBlock;
import com.squareup.kotlinpoet.CodeBlocks;
import com.squareup.kotlinpoet.FileSpec;
import com.squareup.kotlinpoet.KModifier;
import com.squareup.kotlinpoet.ParameterizedTypeName;
import com.squareup.kotlinpoet.PropertySpec;
import com.squareup.kotlinpoet.TypeName;
import com.squareup.kotlinpoet.TypeNames;
import com.squareup.kotlinpoet.TypeSpec;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

final class KotlinWriter implements RulesWriter {

  private final Path outputDir;

  KotlinWriter(Path outputDir) {
    this.outputDir = outputDir;
  }

  @Override public void writeZoneIds(String packageName, String version, Set<String> zoneIds)
      throws IOException {
    TypeSpec typeSpec = TypeSpec.objectBuilder("LazyZoneRules")
        .addModifiers(KModifier.INTERNAL)
        .addProperty(version(version))
        .addProperty(regionId(zoneIds))
        .build();

    if (!Files.exists(outputDir)) {
      Files.createDirectories(outputDir);
    }
    FileSpec.get(packageName, typeSpec)
        .writeTo(outputDir);
  }

  private PropertySpec version(String version) {
    return PropertySpec.builder("VERSION", TypeNames.STRING, KModifier.CONST)
        .initializer("%S", version)
        .build();
  }

  private PropertySpec regionId(Set<String> allRegionIds) {
    CodeBlock.Builder builder = CodeBlock.builder()
        .add("listOf(\n⇥⇥");
    List<CodeBlock> blocks = allRegionIds.stream()
        .map(id -> CodeBlock.of("%S", id))
        .collect(Collectors.toList());

    CodeBlock joined = CodeBlocks.joinToCode(blocks, ",\n");
    builder.add(joined)
        .add("⇤⇤\n)");
    TypeName listType = ParameterizedTypeName.get(TypeNames.LIST, TypeNames.STRING);
    return PropertySpec.builder("REGION_IDS", listType)
        .initializer(builder.build())
        .build();
  }
}

File: ticktock-compiler/src/main/kotlin/dev/zacsweers/ticktock/compiler/KotlinWriter.kt
/*
 * Copyright (C) 2020 Zac Sweers & Gabriel Ittner
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.zacsweers.ticktock.compiler

import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier.CONST
import com.squareup.kotlinpoet.KModifier.INTERNAL
import com.squareup.kotlinpoet.KModifier.OVERRIDE
import com.squareup.kotlinpoet.KModifier.PRIVATE
import com.squareup.kotlinpoet.LIST
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.STRING
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.joinToCode
import dev.zacsweers.ticktock.runtime.ZoneIdsProvider
import java.nio.file.Files
import java.nio.file.Path

internal class KotlinWriter(private val outputDir: Path) : RulesWriter {

  override fun writeZoneIds(
    packageName: String,
    version: String,
    zoneIds: Set<String>
  ) {
    val versionProperty = version(version)
    val zoneIdsProperty = regionId(zoneIds)
    val typeSpec = TypeSpec.objectBuilder("GeneratedZoneIdsProvider")
        .addSuperinterface(ZoneIdsProvider::class)
        .addModifiers(INTERNAL)
        .addProperty(versionProperty)
        .addFunction(FunSpec.builder("getVersionId")
            .addModifiers(OVERRIDE)
            .returns(versionProperty.type)
            .addStatement("return %N", versionProperty)
            .build())
        .addProperty(zoneIdsProperty)
        .addFunction(FunSpec.builder("getZoneIds")
            .addModifiers(OVERRIDE)
            .returns(zoneIdsProperty.type)
            .addStatement("return %N", zoneIdsProperty)
            .build())
        .build()

    Files.createDirectories(outputDir)
    FileSpec.get(packageName, typeSpec)
        .writeTo(outputDir)
  }

  private fun version(version: String): PropertySpec {
    return PropertySpec.builder("VERSION_ID", STRING, PRIVATE, CONST)
        .initializer("%S", version)
        .build()
  }

  private fun regionId(allRegionIds: Set<String>): PropertySpec {
    val blocks = allRegionIds.map { CodeBlock.of("%S", it) }
    val joinedBlocks = blocks.joinToCode(",\n")
    val initializer = CodeBlock.builder()
        .add("listOf(\n⇥⇥")
        .add(joinedBlocks)
        .add("⇤⇤\n)")
        .build()

    val listType = LIST.parameterizedBy(STRING)
    return PropertySpec.builder("ZONE_IDS", listType, PRIVATE)
        .initializer(initializer)
        .build()
  }
}
